src/Permissions.ts

```diff
@@ -0,0 +1,214 @@
+    }
+    return neededSignatures
+  }
+
+  private addSignature (response: Response): number {
+    const signatures = this.signatures.get(response.id)
+    if (signatures === undefined) {
+      this.signatures.set(response.id, new Set(response.from))
+      return 1
+    }
+    if (signatures.has(response.from)) {
+      throw new Error(`${response.from} tried to sign the same request twice`)
+    }
+    signatures.add(response.from)
+    return signatures.size
+  }
+
+  private finishRequest (request: Request): void {
+    if (request.operation === 'merge') {
+      throw new Error('todo')
+    }
+    if (request.operation === 'add') {
+      this.members.set(request.who, 'added')
+    } else {
+      this.members.set(request.who, 'removed')
+    }
+    this.requests.set(request.id, 'finished')
+    this.openRequests.delete(request.id)
+    this.signatures.delete(request.id)
+    const list = this.openRequestsByMember.get(request.from)
+    if (list === undefined) {
+      throw new Error('This may never occur')
+    }
+    if (list.shift() !== request) {
+      throw new Error('This may also never occur')
+    }
+    const entry = list[0]
+    if (entry === undefined) {
+      this.openRequestsByMember.delete(request.from)
+    } else {
+      this.requests.set(entry.id, 'active')
+    }
+  }
+
+  private handleCancel (response: Response, openRequest?: Request): void {
+    if (openRequest !== undefined) {
+      if (openRequest.from !== response.from) {
+        throw new Error(`Member ${response.from} can not cancel the request ${response.id} by ${openRequest.from}.`)
+      }
+      this.requests.set(response.id, 'cancelled')
+      this.openRequests.delete(response.id)
+      this.openRequestsByMember.delete(response.id)
+    }
+    // TODO: Should we throw an error if the request is not active?
+  }
+
+  private handleDeny (response: Response, openRequest?: Request): void {
+    if (openRequest !== undefined) {
+      if (openRequest.from === response.from) {
+        throw new Error(`Member ${response.from} can not deny their own request ${response.id}. Maybe they meant to cancel?`)
+      }
+      this.requests.set(response.id, 'denied')
+      this.openRequests.delete(response.id)
+      this.openRequestsByMember.delete(response.id)
+    }
+    // TODO: Should we throw an error if the request is not active?
+  }
+
+  private handleRequest (request: Request): void {
+    if (request.operation === 'merge') {
+      throw new Error('todo')
+    }
+    if (this.requests.has(request.id)) {
+      throw new Error(`Request ID=${request.id} has already been used.`)
+    }
+    const memberState = this.members.get(request.who)
+    if (request.operation === 'remove') {
+      if (memberState === undefined) {
+        throw new Error(`Cant remove ${request.who} because it is not a member.`)
+      }
+    } else {
+      if (memberState === 'removed') {
+        throw new Error(`Cant add previously removed member ${request.who}`)
```

Comment:
Why can't we re-add a member after removing them? Seems like a legit use case.

Reply:
This was added to make it easier to deal with member messages. If we can re-add members we need to check the messages of that member if it contains messages within a period of time where the member was un-added and ignore those messages while keeping the messages when the member was added. It was also a protection against possibly malicious feeds.

Reply:
So if a person wants to be re-added to the group they need to generate a new member identity and stuff? Seems reasonable.